### PR TITLE
feat: add snp_calc_id_block

### PR DESCRIPTION
**Pull request by shuk777**
Add a snp_calc_id_block to calculate id-block and id-auth, example usage is as follows:

sev-snp-measure --mode snp --vcpus=1 --vcpu-type=EPYC-v4 --ovmf=OVMF.fd --kernel=vmlinuz --initrd=initrd.img --append="console=ttyS0 loglevel=7" --output-format=idblock --idkey=id_key.pem --authorkey=author_key.pem 

The output format:
id-block={base64_encoded},id-auth={base64_encoded}\nid_key_hash: {base64_encoded digest}\nauthor_key: {base64_encoded digest}

### CODE REVIEW

**Comment by dubek** [comment]
Hi @shuk777 , thanks for your contribution to sev-snp-measure. I haven't tried launching VMs with id-block yet, but I understand it's an important use case.

However, looking at the code, it seems that the addition has almost no connection to the measurement calculation; the only connection is that it takes the calculated launch digest and then uses it to build and sign the id-block. So, maybe a cleaner approach (instead of adding an output format) would be to add another utility (it can be in this repo) which is called `snp-create-id-block` (or similar); it will take the expected launch digest as an argument (or from stdin). This way each part is doing one simple thing. Please let me know what you think about this idea.

Other general comments:
1. Missing tests for the new feature.
2. The cryptography imports are a new dependency, probably should be declared in `requirements.txt` and in `setup.cfg`
3. Move the new functions to separate file `id_block.py`
4. In `snp_calc_id_block()`, instead of building the binary struct using `+=` on byte arrays, I suggest using `ctypes.Structure`, as we do in `vmsa.py` or `sev_hashes.py`. Simiarly a `ctypes.Structure` type can be defined for the marshalled EC public key. The struct definitions should be similar to ones that appear in C code or in the SNP spec document.
5. New user-facing changes should be mentioned in the README; perhaps with an example.

Let me know what you think, and thanks again!

**Comment by shuk777** [comment]
Thanks @dubek for your kind comments! Adding a new utility sounds good to me, since putting id-block calculation in sev-snp-measure may cause a confusion for users. 

Already pushed my fix, I'm new to python, so not sure if I'm doing things right. Plz correct me if I made any rookie mistake.

File: sevsnpmeasure/id_block.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    if args.idkey == None or args.authorkey == None:
+        parser.error("missing key files for id block")
+    ld = base64.b64decode(args.measurement)
+    measurement = snp_calc_id_block(ld,args.idkey,args.authorkey)
+
+    print(measurement)
+    return 0
+
+if __name__ == '__main__':
+    sys.exit(main())
```
Comment (dubek) [inline sevsnpmeasure/id_block.py:51]: remove - this module should not be main. We have `snp-create-id-block.py` for that.

File: sevsnpmeasure/id_block.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    parser.add_argument('--idkey', metavar='PATH', help='id private key file')
+    parser.add_argument('--authorkey', metavar='PATH', help='author private key file')
+    args = parser.parse_args()
+
+    if args.idkey == None or args.authorkey == None:
+        parser.error("missing key files for id block")
+    ld = base64.b64decode(args.measurement)
+    measurement = snp_calc_id_block(ld,args.idkey,args.authorkey)
+
+    print(measurement)
```
Comment (dubek) [inline sevsnpmeasure/id_block.py:47]: rename variable -- this is not a measurement.
also add spaces after commas in the argument list.

File: sevsnpmeasure/id_block.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        password=None,  # Change this to a password if your PEM file is encrypted
+        backend=default_backend()
+    )
+    if isinstance(private_key, ec.EllipticCurvePrivateKey):
+        return private_key
+
+    raise ValueError("The provided PEM file does not contain an EC private key.")
+
+def marshal_ec_public_key(priv_key: ec.EllipticCurvePrivateKey) -> bytes:
+    curve = b'\x02\x00\x00\x00'  # P-384
```
Comment (dubek) [inline sevsnpmeasure/id_block.py:143]: unused, remove

**Review by dubek — CHANGES_REQUESTED** [review]
Thank you @shuk777 for modifying the code. It is now better and more decoupled.

There are few technical issues which you can see in the CI checks. Also you can locally run the following commands from the root of the cloned repository and make sure they pass without warnings/errors:

```
make lint
make typecheck
make test
```

And also see some small comments below. 

Feel free to ask if something is not clear.

**Comment by shuk777** [comment]
> Thank you @shuk777 for modifying the code. It is now better and more decoupled.
> 
> There are few technical issues which you can see in the CI checks. Also you can locally run the following commands from the root of the cloned repository and make sure they pass without warnings/errors:
> 
> ```
> make lint
> make typecheck
> make test
> ```
> 
> And also see some small comments below.
> 
> Feel free to ask if something is not clear.

Thank you @dubek for your patience and kind comments. Already pushed new fixes

File: Makefile [hunk 4]
```diff
@@ -4,7 +4,7 @@ test:
 	python3 -m unittest discover --verbose tests/
 
 lint:
-	flake8 . --count --max-complexity=10 --max-line-length=127 --statistics
+	flake8 . --count --max-complexity=10 --max-line-length=152 --statistics
```
Comment (dubek) [inline Makefile:7]: please keep the original value.
if you have a long string, you can break it to two lines; see for example how it's done in test_guest.py:

https://github.com/IBM/sev-snp-measure/blob/63ae739b356cc00bf20fb1e13129ee16f25a4787/tests/test_guest.py#L23-L26

File: setup.cfg [hunk 5]
```diff
@@ -21,6 +21,9 @@ classifiers =
 [options]
 packages = sevsnpmeasure
 python_requires = >=3.7
+install_requires = 
+    cryptography==39.0.1
+    types-cryptography
```
Comment (dubek) [inline setup.cfg:26]: Please also add a `requirements.txt` file at the repo's root dir with the following content:

```
cryptography==39.0.1
types-cryptography
```

I think this will allow the CI jobs to install these prior to running the other tasks (`make typecheck` etc).

File: setup.cfg [hunk 6]
```diff
@@ -21,6 +21,9 @@ classifiers =
...
 
 [options.entry_points]
 console_scripts = sev-snp-measure = sevsnpmeasure.cli:main
```
Comment (dubek) [inline setup.cfg:29]: we need to figure out how to add another console script here (the new `snp-create-id-block` script introduced in this PR).

File: tests/test_guest.py [hunk 7]
```diff
@@ -16,7 +16,7 @@ def test_snp(self):
                 SevMode.SEV_SNP,
                 1,
                 vcpu_types.CPU_SIGS["EPYC-v4"],
-                "tests/fixtures/ovmf_suffix.bin",
+                "fixtures/ovmf_suffix.bin",
```
Comment (dubek) [inline tests/test_guest.py:19]: why is this changed?

**Review by dubek — CHANGES_REQUESTED** [review]
Thanks @shuk777 , it's looking better and getting closer to merge. See some more comments below:

**Comment by dubek** [comment]
Now the tests pass in CI, good!

But the changes in `setup.cfg` were reverted. I think they should stay. And also add another `console_script` (need to find the correct syntax to add two of them). `setup.cfg` affects the wheel that I later generate and publish on pypi (to allow people to use this simply with `pip install sev-snp-measure`).

**Comment by dubek** [comment]
I tried to build a wheel and install it, but it doesn't install the dependencies. Please add this to `setup.cfg` (below the `python_requires` line):

```
install_requires =
    cryptography>=39.0.1
    types-cryptography
```

**Comment by shuk777** [comment]
> I tried to build a wheel and install it, but it doesn't install the dependencies. Please add this to `setup.cfg` (below the `python_requires` line):
> 
> ```
> install_requires =
>     cryptography>=39.0.1
>     types-cryptography
> ```

fixed

**Comment by dubek** [comment]
Merged! Thanks @shuk777 for your contribution.

I hope to cut a new release soon (waiting for some feedback on #12 ).

**Comment by dubek** [comment]
Thanks again @shuk777 . This is now part of release sev-snp-measure 0.0.4 on pypi.
